server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java

```diff
@@ -142,6 +150,66 @@ public void testCancelException() {
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
     }
 
+    public void testAssertCompleteAllowedAllowsConcurrentCompletes() throws InterruptedException {
+        assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
+        try (TestThreadPool tp = new TestThreadPool(getTestName())) {
+            final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
+            final int threads = 4;
+            final AtomicBoolean running = new AtomicBoolean(true);
+            final CountDownLatch startBarrier = new CountDownLatch(threads + 1);
+            final CompletionService<?> cs = new ExecutorCompletionService<>(tp.generic());
+            final List<Future<?>> futures = new ArrayList<>();
+            for (int i = 0; i < threads; i++) {
+                futures.add(cs.submit(new FutureCompleter(i, startBarrier, futureReference, running), null));
+            }
+            startBarrier.countDown();
+            safeAwait(startBarrier);
+            cs.poll(500, TimeUnit.MILLISECONDS);
+            running.set(false);
+            futures.forEach(ESTestCase::safeGet);
+        }
+    }
+
+    private static class FutureCompleter implements Runnable {
+        private final int number;
+        private final AtomicReference<PlainActionFuture<?>> future;
+        private final AtomicBoolean running;
+        private final CountDownLatch startBarrier;
+        private volatile boolean failed = false;
+
+        private FutureCompleter(
+            int number,
+            CountDownLatch startBarrier,
+            AtomicReference<PlainActionFuture<?>> future,
+            AtomicBoolean running
+        ) {
+            this.number = number;
+            this.future = future;
+            this.running = running;
+            this.startBarrier = startBarrier;
+        }
+
+        @Override
+        public void run() {
+            try {
+                startBarrier.countDown();
+                safeAwait(startBarrier);
+                while (running.get()) {
+                    if (number == 0) {
+                        future.set(new PlainActionFuture<>());
+                    }
+                    future.get().onResponse(null);
+                }
```

Comment:
The issue only manifests in the moment the future is first resolved, this test reproduces fairly quickly when we do the above.